src/algorithms/fardetectors/FarDetectorLinearTracking.cc

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023, Simon Gardner
+
+#include <edm4hep/Vector2f.h>
+#include <edm4hep/Vector3f.h>
+#include <edm4eic/Cov2f.h>
+#include <edm4eic/Cov3f.h>
+#include <edm4eic/TrackSegmentCollection.h>
+#include <edm4hep/utils/vector_utils.h>
+#include <edm4eic/vector_utils.h>
+#include <edm4hep/TrackerHitCollection.h>
+
+#include "FarDetectorLinearTracking.h"
+#include "services/log/Log_service.h"
+#include "extensions/spdlog/SpdlogExtensions.h"
+#include <iterator>
+#include <algorithm>
+#include <map>
+
+
+
+namespace eicrecon {
+
+    void FarDetectorLinearTracking::init(std::shared_ptr<spdlog::logger>& logger) {
+
+      m_log      = logger;
+
+      // For changing how strongly each layer hit is in contributing to the fit
+      m_layerWeights = Eigen::VectorXd::Constant(m_cfg.n_layer,1);
+
+    }
+
+    void FarDetectorLinearTracking::process(
+        const FarDetectorLinearTracking::Input& input,
+        const FarDetectorLinearTracking::Output& output) const {
+
+        const auto [inputhits] = input;
+        auto [outputTracks] = output;
+
+        // Check the number of input collections is correct
+        int nCollections = inputhits.size();
+        if(nCollections!=m_cfg.n_layer){
+          m_log->error("Wrong number of input collections passed to algorithm");
+          return;
+        }
+
+        // Check there aren't too many hits in any layer to handle
+        // Temporary limit of number of hits per layer before Kalman filtering/GNN implemented
+        // TODO - Implement more sensible solution
+        for(const auto& layerHits: inputhits){
+          if((*layerHits).size()>m_cfg.layer_hits_max){
+            m_log->info("Too many hits in layer");
+            return;
+          }
+        }
+
+        // Create a matrix to store the hit positions
+        Eigen::MatrixXd hitMatrix(3,m_cfg.n_layer);
+        // Loop over all combinations of hits fitting a track to all layers
+        makeHitCombination(m_cfg.n_layer-1,&hitMatrix,inputhits,outputTracks);
+
+    }
+
+
+    void FarDetectorLinearTracking::makeHitCombination(int level,
+                                                       Eigen::MatrixXd* hitMatrix,
+                                                       const std::vector<gsl::not_null<const edm4hep::TrackerHitCollection*>>& hits,
+                                                       gsl::not_null<edm4eic::TrackSegmentCollection*> outputTracks ) const {
+
+      // Iterate over hits in this layer
+      for(auto hit : (*hits[level])){
+        auto pos = hit.getPosition();
+        hitMatrix->col(level) << pos.x, pos.y, pos.z;
+
+        if(level>0){
+          makeHitCombination(level-1,
+                             hitMatrix,
+                             hits,
+                             outputTracks);
+        }
+        else{
+          checkHitCombination(hitMatrix,outputTracks);
```

Comment:
Move this to process. Does this even need to be recursive?

Reply:
This is only called on the inner most recursion, so unless all possible hit matrix combinations are stored, then looped over separately in process. There are probably some nice std functions which would iterate over the combinations but likely have to set all components of the matrix on each iteration rather than just the current recursion step. The reason this is recursive rather than a set number of nested loops was to allow for various numbers of layers to be included. I'm sure acts has nice methods to handle all of this again itself.